dali/pipeline/operators/reader/reader_op.h

```diff
@@ -120,22 +120,23 @@ class DataReader : public Operator<Backend> {
   }
 
   // CPUBackend operators
-  void Run(SampleWorkspace* ws) override {
+  void Run(HostWorkspace* ws) override {
     // If necessary start prefetching thread and wait for a consumable batch
     StartPrefetchThread();
     ConsumerWait();
 
-    // consume sample
+    // consume batch
     TimeRange tr("DataReader::Run #" + to_string(curr_batch_consumer_), TimeRange::kViolet);
+
+    // This is synchronous call for CPU Backend
     Operator<Backend>::Run(ws);
-    const auto data_idx = ws->data_idx();
-    loader_->RecycleTensor(MoveSample(data_idx));
-    auto curr_sample_id = samples_processed_.fetch_add(1);
-    // if we've processed the whole batch, notify it
-    if (curr_sample_id == Operator<Backend>::batch_size_ - 1) {
-      samples_processed_ = 0;
-      ConsumerAdvanceQueue();
+
```

Comment:
just thinking. Is there a performance hit for doing this sequentially?

Reply:
The `Operator<Backend>::Run(ws);` uses thread pool internally, RecycleTensor is done the same way as for GPU. It may still affect the pref, but shouldn't be a big change.
We also removed the atomic counting of runs, so that might help a bit.